identify/src/lib.rs

```diff
@@ -0,0 +1,289 @@
+        service.send_message(session.id, self.id, data);
+        let remote_addr =
+            multiaddr_to_socketaddr(&session.address).expect("Can not get remote address");
+        let data = IdentifyMessage::ObservedAddr(remote_addr).encode();
+        service.send_message(session.id, self.id, data);
+    }
+
+    fn disconnected(&mut self, _service: &mut ServiceContext, session: &SessionContext) {
+        let info = self
+            .remote_infos
+            .remove(&session.id)
+            .expect("RemoteInfo must exists");
+        trace!("IdentifyProtocol disconnected from {:?}", info.peer_id);
+    }
+
+    fn received(
+        &mut self,
+        service: &mut ServiceContext,
+        session: &SessionContext,
+        data: bytes::Bytes,
+    ) {
+        let info = self
+            .remote_infos
+            .get_mut(&session.id)
+            .expect("RemoteInfo must exists");
+        match IdentifyMessage::decode(&data) {
+            Some(IdentifyMessage::ListenAddrs(addrs)) => {
+                if info.listen_addrs.is_some() {
+                    debug!("remote({:?}) repeat send observed address", info.peer_id);
+                    if self
+                        .addr_mgr
+                        .misbehave(session.address.clone(), Misbehavior::DuplicateListenAddrs)
+                        .is_disconnect()
+                    {
+                        service.disconnect(session.id);
+                    }
+                } else if addrs.len() > MAX_ADDRS {
+                    if self
+                        .addr_mgr
+                        .misbehave(
+                            session.address.clone(),
+                            Misbehavior::TooManyAddresses(addrs.len()),
+                        )
+                        .is_disconnect()
+                    {
+                        service.disconnect(session.id);
+                    }
+                } else {
+                    trace!("received listen addresses: {:?}", addrs);
+                    self.addr_mgr
+                        .add_listen_addrs(session.address.clone(), addrs.clone());
+                    info.listen_addrs = Some(addrs);
+                }
+            }
+            Some(IdentifyMessage::ObservedAddr(addr)) => {
+                if info.observed_addr.is_some() {
+                    debug!("remote({:?}) repeat send listen addresses", info.peer_id);
+                    if self
+                        .addr_mgr
+                        .misbehave(session.address.clone(), Misbehavior::DuplicateObservedAddr)
+                        .is_disconnect()
+                    {
+                        service.disconnect(session.id);
+                    }
+                } else {
+                    trace!("received observed address: {}", addr);
+                    info.observed_addr = Some(addr);
+                    // TODO how can we trust this address?
```

Comment:
should use a HashMap<Multiaddr, Score> to track how many times we received the same address from a remote peer, we can trust an address by its score.

Or add  `addr_score` function to the trait, CKB network can impl it with peerstore.

Reply:
How about use `add_observed_addr` API to get feedback?

Reply:
Seems OK to me.